src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,557 @@
+        """
+        Populates mass profiles for every tree defined in the ensemble.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+
+        boolean: boolean, True or False
+            True to update reference mass, False to update latest mass
+        """
+        for i in range(self.n_estimators):
+            self.ensemble[i].update_mass(X, boolean)
+
+    def update_models(self):
+        """
+        Updates the mass profile of every tree in the ensemble.
+
+        """
+        for i in range(self.n_estimators):
+            tree = self.ensemble[i]
+            tree.update_model(tree.root)
+
+    def set_is_learning_phase_on(self, boolean):
+        """
+        Sets learning phase in each tree defined in the ensemble.
+
+        Parameters
+        ----------
+        boolean: Boolean
+
+        """
+        self.is_learning_phase_on = boolean
+        for i in range(self.n_estimators):
+            self.ensemble[i].is_learning_phase_on = boolean
+
+
+class HalfSpaceTree:
+
+    def __init__(self, max_depth, n_features, size_limit, min_values, max_values, tree_random_state=None):
+        """
+        Half Space Tree
+
+        Parameters
+        ----------
+        n_features: int, required
+            The dimensionality of the stream.
+
+        max_depth: int, optional (default=15)
+            The maximum depth of the trees in the ensemble.
+            'maxDepth' in the original paper.
+
+        size_limit: int, optional (default=50)
+            The minimum mass required in a node (as a fraction of the window size) to calculate the anomaly score.
+            'sizeLimit' in the original paper.
+
+        min_values: Array of floats
+            Minimum work range for every dimension.
+
+        max_values: Array of floats
+            Maximum work range for every dimension.
+
+        tree_random_state: int (Default: None)
+            Seed for random generation of tree.
+        """
+        super().__init__()
+        self.max_depth = max_depth
+        self.n_features = n_features
+        self.size_limit = size_limit
+        self.is_learning_phase_on = True
+        self.root = self.build_tree(min_values, max_values)
+        random.seed(tree_random_state)
+
+    def predict_proba(self, X, max_score):
+        """ Predicts probabilities of all label of the X instance(s)
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (n_samples, n_features)
+            Samples for which we want to predict the labels.
+
+        max_score: float
+            The maximum score of an instance could have in the tree.
+
+        Returns
+        -------
+        numpy.array
+            Predicted the probabilities of all the labels for all instances in X.
+
+        """
+        r, _ = get_dimensions(X)
+        predictions = []
+        for i in range(r):
+            votes = self.get_votes_for_instance(X[i], max_score).copy()
```

Comment:
Please use `copy.deepcopy` to protect the internal votes. We have had some issues due to `.copy()` not working as expected.

```python
            votes = copy.deepcopy(self.get_votes_for_instance(X[i], max_score))
```